Inferences/Induction.hpp

```diff
@@ -42,11 +42,19 @@ using namespace Saturation;
 class TermReplacement : public TermTransformer {
 
 public:
-  TermReplacement(Term* o, TermList r) : _o(o), _r(r) {} 
-  virtual TermList transformSubterm(TermList trm);
+  TermReplacement(Term* o, TermList r, bool replaceSkolems = false, unsigned* var = nullptr)
+    : _o(o), _r(r), _v(var), _replaceSkolems(replaceSkolems)
+  {
+    ASS(!replaceSkolems || var);
+  }
+  TermList transformSubterm(TermList trm) override;
+  VList* getUsedVars() const;
 private:
   Term* _o;
   TermList _r;
+  unsigned* _v;
+  bool _replaceSkolems;
+  DHMap<Term*, unsigned> _tv;
```

Comment:
Also, do we use `TermReplacement` in any other context? If so, from a pay-for-what-you-use perspective it could be nice to implement `class SkolemSquashingTermReplacement: public TermReplacement` as a subclass, which also eliminates `_replaceSkolems` via dynamic dispatch. I leave whether to do this (and naming the class) up to you!

Reply:
Good points! `TermReplacement` is only used within `Induction` and the feature of this PR should be implemented for integer induction as well (I just didn't want to mess up that part of the code too plus it's a bit more complicated there because the Skolems can appear in the bounds).

Maybe a more general intermediate solution would be to specialize as you suggested and add some helper function that hides the ugly parts.

Tbh, I don't even see now why `TermReplacement` was implemented instead of using more general methods like `EqHelper::replace`. Anyways, the main idea of putting the Skolem replacement inside `TermReplacement` was to do simultaneous replacement instead of going over the literal `n+1` times and introducing a lot of unused shared terms. This is of course orthogonal to your suggestion but my point is that maybe there is some already existing solution I can't see laying around somewhere in the code which would better fit this functionality than the original `TermReplacement`.

Reply:
OK, then I leave it up to you - I wouldn't worry about possible duplication too much, there's a fair amount of it already and often the "general" function doesn't quite work for what you need. Thanks for explaining. :-)